Keep documentation comments in structures with no other items (MPR#7701 cont.)

lpw25 opened the pull request:
This stops us from dropping documentation comments in structures and signatures that have no other items in them. It also adds some more docstring tests and fixes some typos in the ocamldoc documentation.

This is a more complete fix than #1562, which only handled the case where the signature in question was an entire mli file.

lpw25 committed:
Don't drop doc comments in empty signatures

lpw25 committed:
Add more doc comment tests

lpw25 committed:
Fix syntax errors in ocamldoc examples

tarptaeya commented on testsuite/tests/docstrings/docstrings.ml:
```diff
@@ -0,0 +1,575 @@
+    (** The comment for variable x. *)
+    val mutable x : int
+
+    (** The commend for method m. *)
```
This must be comment, not commend

tarptaeya reviewed (commented):
There is a typo, please correct it

gasche commented on manual/manual/cmds/ocamldoc.etex:
```diff
@@ -494,8 +494,8 @@ type weather2 =
 (** The comment for type my_record *)
 type my_record = {
-    val foo : int ;    (** Comment for field foo *)
-    val bar : string ; (** Comment for field bar *)
```
This was grossly invalid. Is it possible that using one of the `{caml_example}` environments would have caught the mistake earlier? If applicable in this case, I would recommend using them as part of this PR -- if we fix, let's fix long-term. cc @Octachron who may be able to comment on applicability.

Octachron replied:
In this case, the code example is a signature; this would require a small tweak to `caml_tex2.ml` to add an option to wrap code example inside a module type declaration.

gasche commented on testsuite/tests/docstrings/docstrings.ml:
```diff
@@ -0,0 +1,575 @@
+module Manual :
+  sig
+    [@@@ocaml.text
+      " Special comments can be placed between elements and are kept\n      by the OCamldoc tool, but are not associated to any element.\n      @-tags in these comments are ignored."]
```
is there a way to use line breaks in the `expect` result here that is portable and correct, and increases readability of the test file?

lpw25 replied:
Without changing how `-dsource` prints things, any such change would be undone when promoting the result of the test.

gasche replied:
So, alternative question: is there a way for `-dsource` to print this in a nicer way? (This would be a separate PR.)

gasche replied:
I think that 

```
"  foo
  bar"
```
is not portable due to \n vs. \r\n concerns,

```
"  foo\n\
  bar"
```

does not work as spaces at the beginning of the line are ignored, but that one could use


```
"  foo\n  \
   bar"
```

pushing indentation of a line on the previous line.

Octachron commented on testsuite/tests/docstrings/docstrings.ml:
```diff
@@ -0,0 +1,575 @@
+
+module M = struct
+
+  (** foo1 *)
```
It would be nice to capture the warning 50 raised by those unattached documentation comments, but this would require patching `expect_text.ml` . However, it could be nice to explicit this point here 
`(** foo1: this comment is unattached*)` .

Octachron commented on testsuite/tests/docstrings/docstrings.ml:
```diff
@@ -0,0 +1,575 @@
+                                                           " The comment for type my_record "]
+    class my_class =
+      object
+        inherit  cl[@@ocaml.doc
```
Would it makes sense to permute the definition of `my_class` and `my_class_type` and inherit `my_class_type` here? After fixing the type of the method `m`, it would avoid the unbound class error later.

lpw25 replied:
I don't think you can inherit from a class type there.

Octachron replied:
You are right of course, I was thinking of the signature part.

Octachron commented on testsuite/tests/docstrings/docstrings.ml:
```diff
@@ -0,0 +1,575 @@
+|}]
+
+(***********************************************************************)
+(* Empty doc comments (GPR#548) *)
```
By pure curiousity, is there a reason beyond personal preference for having a single file with multiple parts rather than multiple files? And with a single file, would it not be better to merge the `docstrings` folder with the `parsetree` one, in order to reduce the number of tests folder in the test suite?

lpw25 replied:
Nothing beyond personal preference. I tend to like single test files that act as a form of documentation for how a particular feature behaves. Since I wanted to switch the "empty" tests to use a `-dsource` expect test anyway, I figured I might as well put them with the others.

> And with a single file, would it not be better to merge the docstrings folder with the parsetree one, in order to reduce the number of tests folder in the test suite?

Yeah, that's probably a good idea.

Octachron commented on testsuite/tests/docstrings/docstrings.ml:
```diff
@@ -0,0 +1,575 @@
+
+  type s
+  (** bar1 *)
+  (** bar2 *)
```
Same comment as above: `(** bar2: this comment is unattached *)`?

Octachron reviewed (approved):
The core changes implement the right fixes. I have some minor existential interrogations on the docstring test but nothing that really matters.

lpw25 committed:
Merge the docstring tests into parsing directory

lpw25 commented:
Ok I moved the docstring tests into `docstring.ml` in `tests/parsing` -- which already existed and had yet another docstring test. I've also updated the comments as suggested. I haven't fixed the example code to type-check because it's not needed and I would prefer to keep the code close to the example in the ocamldoc manual.

trefis commented:
@Octachron : assuming a Changes entry was added, are you now satisfied with the state of this PR and can it be merged?

Also, I'm wondering if this bugfix could/should be backport to the 4.07 branch. Any opinion?

Octachron commented:
Yes, I am totally satisfied with the current state. Sorry for not being explicit. Backporting seems fine but not urgent to me since there is a workaround for the issue.

pmetzger commented:
A pure side comment. It would be nice for people reading the pull requests if the title of something like this was:

`Fix MPR#7701: "Doc comment dropped from empty .mli file"`

which make it easy for a casual reader to remember if is was a Mantis issue they are following.
